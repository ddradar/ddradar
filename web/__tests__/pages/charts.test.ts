import { testSongData as song } from '@ddradar/core/__tests__/data'
import Oruga from '@oruga-ui/oruga-next'
import { bulmaConfig } from '@oruga-ui/theme-bulma'
import { RouterLinkStub } from '@vue/test-utils'
import { describe, expect, test, vi } from 'vitest'
import { ref } from 'vue'

import { useFetch, useRoute } from '#app'
import { mountAsync } from '~/__tests__/test-utils'
import Page from '~/pages/charts.vue'

vi.mock('#app')

describe('Page /charts', () => {
  const query = { style: '1', level: '16' }
  const charts = song.charts.map(c => ({
    id: song.id,
    name: song.name,
    series: song.series,
    playStyle: c.playStyle,
    difficulty: c.difficulty,
    level: c.level,
  }))

  describe('snapshot tests', () => {
    test('{ isLoading: true } renders loading state', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useRoute).mockReturnValue({ query } as any)
      vi.mocked(useFetch).mockResolvedValue({
        pending: ref(true),
        data: ref([]),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig]],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
    test('{ isLoading: false, songs: <Data> } renders course list', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useRoute).mockReturnValue({ query } as any)
      vi.mocked(useFetch).mockResolvedValue({
        pending: ref(false),
        data: ref(charts),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig]],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
    test('{ isLoading: false, songs: [] } renders no data', async () => {
      // Arrange
      /* eslint-disable @typescript-eslint/no-explicit-any */
      vi.mocked(useRoute).mockReturnValue({ query } as any)
      vi.mocked(useFetch).mockResolvedValue({
        pending: ref(false),
        data: ref([]),
      } as any)
      /* eslint-enable @typescript-eslint/no-explicit-any */

      // Act
      const wrapper = await mountAsync(Page, {
        global: {
          plugins: [[Oruga, bulmaConfig]],
          stubs: { NuxtLink: RouterLinkStub },
        },
      })

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  test.each([
    ['1', '6', 'SINGLE 6'],
    ['2', '18', 'DOUBLE 18'],
  ])('?style=%s&level=%s renders "%s" title', async (style, level, title) => {
    // Arrange
    /* eslint-disable @typescript-eslint/no-explicit-any */
    vi.mocked(useRoute).mockReturnValue({ query: { style, level } } as any)
    vi.mocked(useFetch).mockResolvedValue({
      pending: ref(true),
      data: ref([]),
    } as any)
    /* eslint-enable @typescript-eslint/no-explicit-any */

    // Act
    const wrapper = await mountAsync(Page, {
      global: {
        plugins: [[Oruga, bulmaConfig]],
        stubs: { NuxtLink: RouterLinkStub },
      },
    })

    // Assert
    expect(wrapper.find('h1').text()).toBe(title)
  })
})
