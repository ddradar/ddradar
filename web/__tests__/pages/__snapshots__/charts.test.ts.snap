// Vitest Snapshot v1

exports[`Page /charts > snapshot tests > { isLoading: false, songs: <Data> } renders course list 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    SINGLE 16
  </h1>
  <div
    class="buttons"
  >
    <a
      class="button is-success"
    >
       プレースタイルを切り替える 
    </a>
  </div>
  <div
    class="buttons"
  >
    
    <a
      class="button is-info"
    >
      1
    </a>
    <a
      class="button is-info"
    >
      2
    </a>
    <a
      class="button is-info"
    >
      3
    </a>
    <a
      class="button is-info"
    >
      4
    </a>
    <a
      class="button is-info"
    >
      5
    </a>
    <a
      class="button is-info"
    >
      6
    </a>
    <a
      class="button is-info"
    >
      7
    </a>
    <a
      class="button is-info"
    >
      8
    </a>
    <a
      class="button is-info"
    >
      9
    </a>
    <a
      class="button is-info"
    >
      10
    </a>
    <a
      class="button is-info"
    >
      11
    </a>
    <a
      class="button is-info"
    >
      12
    </a>
    <a
      class="button is-info"
    >
      13
    </a>
    <a
      class="button is-info"
    >
      14
    </a>
    <a
      class="button is-info"
    >
      15
    </a>
    <a
      class="button is-info is-outlined"
      disabled="true"
    >
      16
    </a>
    <a
      class="button is-info"
    >
      17
    </a>
    <a
      class="button is-info"
    >
      18
    </a>
    <a
      class="button is-info"
    >
      19
    </a>
    
  </div>
  <div
    class="b-table"
  >
    <div
      style="display: none;"
    >
      
      <span
        data-id="1"
      >
        Series
      </span>
      <span
        data-id="2"
      >
        Name
      </span>
      <span
        data-id="3"
      >
        Difficulty
      </span>
      <span
        data-id="4"
      >
        Level
      </span>
      
    </div>
    <!--v-if-->
    <!--v-if-->
    <div
      class="table-wrapper"
    >
      <table
        class="table is-striped"
        tabindex="false"
      >
        <!--v-if-->
        <thead>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
              draggable="false"
            >
              <span>
                Series 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Name 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Difficulty 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Level 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            
            <!--v-if-->
          </tr>
          <!--v-if-->
          <!--v-if-->
        </thead>
        <tbody>
          
          
          <tr
            class=""
            draggable="false"
          >
            <!--v-if-->
            <!--v-if-->
            
            <td
              class=""
              data-label="Series"
              scoped=""
            >
              1st
            </td>
            <td
              class=""
              data-label="Name"
              scoped=""
            >
              <a>
                PARANOiA
              </a>
            </td>
            <td
              class=""
              data-label="Difficulty"
              scoped=""
            >
              <span
                class="tag is-beginner"
              >
                BEGINNER
              </span>
            </td>
            <td
              class=""
              data-label="Level"
              scoped=""
            >
              4
            </td>
            
            <!--v-if-->
          </tr>
          <transition-stub>
            <!--v-if-->
          </transition-stub>
          <!--v-if-->
          
          
          <tr
            class=""
            draggable="false"
          >
            <!--v-if-->
            <!--v-if-->
            
            <td
              class=""
              data-label="Series"
              scoped=""
            >
              1st
            </td>
            <td
              class=""
              data-label="Name"
              scoped=""
            >
              <a>
                PARANOiA
              </a>
            </td>
            <td
              class=""
              data-label="Difficulty"
              scoped=""
            >
              <span
                class="tag is-basic"
              >
                BASIC
              </span>
            </td>
            <td
              class=""
              data-label="Level"
              scoped=""
            >
              8
            </td>
            
            <!--v-if-->
          </tr>
          <transition-stub>
            <!--v-if-->
          </transition-stub>
          <!--v-if-->
          
          
          <!--v-if-->
        </tbody>
        <!--v-if-->
      </table>
      
      <transition-stub>
        <!--v-if-->
      </transition-stub>
      
    </div>
    
    <div
      class="level"
    >
      <div>
        
        
      </div>
      <div>
        <div>
          <nav
            class="pagination"
          >
            <a
              aria-current="false"
              class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-left mdi-24px"
                />
              </span>
              
            </a>
            <a
              aria-current="false"
              class="pagination-link pagination-next is-disabled pagination-next is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-right mdi-24px"
                />
              </span>
              
            </a>
            <ul
              class="pagination-list"
            >
              <!--First-->
              <!--v-if-->
              <!--v-if-->
              <!--Pages-->
              
              <li>
                <a
                  aria-current="true"
                  class="pagination-link is-current"
                  href="#"
                  role="button"
                >
                  
                  1
                  
                </a>
              </li>
              
              <!--Last-->
              <!--v-if-->
              <!--v-if-->
            </ul>
          </nav>
        </div>
      </div>
    </div>
    
  </div>
</section>
`;

exports[`Page /charts > snapshot tests > { isLoading: false, songs: [] } renders no data 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    SINGLE 16
  </h1>
  <div
    class="buttons"
  >
    <a
      class="button is-success"
    >
       プレースタイルを切り替える 
    </a>
  </div>
  <div
    class="buttons"
  >
    
    <a
      class="button is-info"
    >
      1
    </a>
    <a
      class="button is-info"
    >
      2
    </a>
    <a
      class="button is-info"
    >
      3
    </a>
    <a
      class="button is-info"
    >
      4
    </a>
    <a
      class="button is-info"
    >
      5
    </a>
    <a
      class="button is-info"
    >
      6
    </a>
    <a
      class="button is-info"
    >
      7
    </a>
    <a
      class="button is-info"
    >
      8
    </a>
    <a
      class="button is-info"
    >
      9
    </a>
    <a
      class="button is-info"
    >
      10
    </a>
    <a
      class="button is-info"
    >
      11
    </a>
    <a
      class="button is-info"
    >
      12
    </a>
    <a
      class="button is-info"
    >
      13
    </a>
    <a
      class="button is-info"
    >
      14
    </a>
    <a
      class="button is-info"
    >
      15
    </a>
    <a
      class="button is-info is-outlined"
      disabled="true"
    >
      16
    </a>
    <a
      class="button is-info"
    >
      17
    </a>
    <a
      class="button is-info"
    >
      18
    </a>
    <a
      class="button is-info"
    >
      19
    </a>
    
  </div>
  <div
    class="b-table"
  >
    <div
      style="display: none;"
    >
      
      <span
        data-id="1"
      >
        Series
      </span>
      <span
        data-id="2"
      >
        Name
      </span>
      <span
        data-id="3"
      >
        Difficulty
      </span>
      <span
        data-id="4"
      >
        Level
      </span>
      
    </div>
    <!--v-if-->
    <!--v-if-->
    <div
      class="table-wrapper"
    >
      <table
        class="table is-striped is-empty"
        tabindex="false"
      >
        <!--v-if-->
        <thead>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
              draggable="false"
            >
              <span>
                Series 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Name 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Difficulty 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Level 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            
            <!--v-if-->
          </tr>
          <!--v-if-->
          <!--v-if-->
        </thead>
        <tbody>
          
          
          <tr>
            <td
              colspan="4"
            >
              
              <section
                class="section"
              >
                <div
                  class="content has-text-grey has-text-centered"
                >
                  <p>
                    No Data
                  </p>
                </div>
              </section>
              
            </td>
          </tr>
        </tbody>
        <!--v-if-->
      </table>
      
      <transition-stub>
        <!--v-if-->
      </transition-stub>
      
    </div>
    
    <div
      class="level"
    >
      <div>
        
        
      </div>
      <div>
        <div>
          <nav
            class="pagination"
          >
            <a
              aria-current="false"
              class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-left mdi-24px"
                />
              </span>
              
            </a>
            <a
              aria-current="false"
              class="pagination-link pagination-next is-disabled pagination-next is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-right mdi-24px"
                />
              </span>
              
            </a>
            <ul
              class="pagination-list"
            >
              <!--First-->
              <!--v-if-->
              <!--v-if-->
              <!--Pages-->
              
              
              <!--Last-->
              <!--v-if-->
              <!--v-if-->
            </ul>
          </nav>
        </div>
      </div>
    </div>
    
  </div>
</section>
`;

exports[`Page /charts > snapshot tests > { isLoading: true } renders loading state 1`] = `
<section
  class="section"
>
  <h1
    class="title"
  >
    SINGLE 16
  </h1>
  <div
    class="buttons"
  >
    <a
      class="button is-success"
    >
       プレースタイルを切り替える 
    </a>
  </div>
  <div
    class="buttons"
  >
    
    <a
      class="button is-info"
    >
      1
    </a>
    <a
      class="button is-info"
    >
      2
    </a>
    <a
      class="button is-info"
    >
      3
    </a>
    <a
      class="button is-info"
    >
      4
    </a>
    <a
      class="button is-info"
    >
      5
    </a>
    <a
      class="button is-info"
    >
      6
    </a>
    <a
      class="button is-info"
    >
      7
    </a>
    <a
      class="button is-info"
    >
      8
    </a>
    <a
      class="button is-info"
    >
      9
    </a>
    <a
      class="button is-info"
    >
      10
    </a>
    <a
      class="button is-info"
    >
      11
    </a>
    <a
      class="button is-info"
    >
      12
    </a>
    <a
      class="button is-info"
    >
      13
    </a>
    <a
      class="button is-info"
    >
      14
    </a>
    <a
      class="button is-info"
    >
      15
    </a>
    <a
      class="button is-info is-outlined"
      disabled="true"
    >
      16
    </a>
    <a
      class="button is-info"
    >
      17
    </a>
    <a
      class="button is-info"
    >
      18
    </a>
    <a
      class="button is-info"
    >
      19
    </a>
    
  </div>
  <div
    class="b-table"
  >
    <div
      style="display: none;"
    >
      
      <span
        data-id="1"
      >
        Series
      </span>
      <span
        data-id="2"
      >
        Name
      </span>
      <span
        data-id="3"
      >
        Difficulty
      </span>
      <span
        data-id="4"
      >
        Level
      </span>
      
    </div>
    <!--v-if-->
    <!--v-if-->
    <div
      class="table-wrapper"
    >
      <table
        class="table is-striped is-empty"
        tabindex="false"
      >
        <!--v-if-->
        <thead>
          <tr>
            <!--v-if-->
            <!--v-if-->
            
            <th
              class=""
              draggable="false"
            >
              <span>
                Series 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Name 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Difficulty 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            <th
              class=""
              draggable="false"
            >
              <span>
                Level 
                <span
                  class="th-wrap sort-icon"
                  style="display: none;"
                >
                  <span
                    class="icon is-small"
                  >
                    <i
                      class="mdi mdi-arrow-up"
                    />
                  </span>
                </span>
              </span>
            </th>
            
            <!--v-if-->
          </tr>
          <!--v-if-->
          <!--v-if-->
        </thead>
        <tbody>
          
          
          <tr>
            <td
              colspan="4"
            >
              
              <section
                class="section"
              >
                <div
                  class="b-skeleton is-animated"
                >
                  <div
                    class="b-skeleton-item is-rounded"
                  />
                  <div
                    class="b-skeleton-item is-rounded"
                  />
                  <div
                    class="b-skeleton-item is-rounded"
                  />
                </div>
              </section>
              
            </td>
          </tr>
        </tbody>
        <!--v-if-->
      </table>
      
      <transition-stub>
        <div
          class="o-load loading"
        >
          <div
            class="o-load__overlay loading-overlay"
          />
          
          <span
            class="icon is-spin is-medium o-load__icon icon"
          >
            <i
              class="mdi mdi-loading mdi-36px"
            />
          </span>
          
        </div>
      </transition-stub>
      
    </div>
    
    <div
      class="level"
    >
      <div>
        
        
      </div>
      <div>
        <div>
          <nav
            class="pagination"
          >
            <a
              aria-current="false"
              class="pagination-link pagination-previous is-disabled pagination-previous is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-left mdi-24px"
                />
              </span>
              
            </a>
            <a
              aria-current="false"
              class="pagination-link pagination-next is-disabled pagination-next is-disabled"
              href="#"
              role="button"
            >
              
              <span
                aria-hidden="true"
                class="icon"
              >
                <i
                  class="mdi mdi-chevron-right mdi-24px"
                />
              </span>
              
            </a>
            <ul
              class="pagination-list"
            >
              <!--First-->
              <!--v-if-->
              <!--v-if-->
              <!--Pages-->
              
              
              <!--Last-->
              <!--v-if-->
              <!--v-if-->
            </ul>
          </nav>
        </div>
      </div>
    </div>
    
  </div>
</section>
`;
